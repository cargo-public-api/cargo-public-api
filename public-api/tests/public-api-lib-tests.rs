// deny in CI, only warn here
#![warn(clippy::all)]

use std::{fs, io::Write, path::PathBuf};

use public_api::Error;

use pretty_assertions::assert_eq;

use tempfile::{NamedTempFile, TempDir, tempdir};

mod common;
use common::{
    builder_for_crate, rustdoc_json_path_for_crate, rustdoc_json_path_for_temp_crate,
    simplified_builder_for_crate,
};

#[test]
fn public_api() -> Result<(), Box<dyn std::error::Error>> {
    public_api_for_manifest("public-api", "Cargo.toml")
}

// To avoid circular workspace dependencies we test the API surface of this
// crate as well.
#[test]
fn public_api_for_rustup_toolchain() -> Result<(), Box<dyn std::error::Error>> {
    public_api_for_manifest("rustup-toolchain", "../rustup-toolchain/Cargo.toml")
}

// To avoid circular workspace dependencies we test the API surface of this
// crate as well.
#[test]
fn public_api_for_rustdoc_json() -> Result<(), Box<dyn std::error::Error>> {
    public_api_for_manifest("rustdoc-json", "../rustdoc-json/Cargo.toml")
}

fn public_api_for_manifest(
    snapshot_name: &str,
    manifest_path: impl AsRef<std::path::Path>,
) -> Result<(), Box<dyn std::error::Error>> {
    let rustdoc_json = rustdoc_json::Builder::default()
        .toolchain("nightly")
        .manifest_path(&manifest_path)
        .build()?;

    let public_api = public_api::Builder::from_rustdoc_json(rustdoc_json).build()?;

    assert_or_bless::assert_eq_or_bless_if(
        public_api.to_string(),
        format!("{snapshot_name}.txt"),
        std::env::var("PUBLIC_API_BLESS")
            .map_or(false, |s| s == "1" || s == "yes" || s == "true"),
    );

    Ok(())
}

#[test]
fn not_simplified() {
    // Create independent build dir so all tests can run in parallel
    let build_dir = tempdir().unwrap();

    assert_public_api(
        builder_for_crate("../test-apis/example_api-v0.2.0", &build_dir),
        "example_api-v0.2.0-not-simplified",
    );
}

#[test]
fn simplified_without_auto_derived_impls() {
    // Create independent build dir so all tests can run in parallel
    let build_dir = tempdir().unwrap();

    assert_public_api(
        builder_for_crate("../test-apis/example_api-v0.2.0", &build_dir)
            .omit_auto_derived_impls(true)
            .omit_blanket_impls(true)
            .omit_auto_trait_impls(true),
        "example_api-v0.2.0-simplified_without_auto_derived_impls",
    );
}

#[test]
fn omit_blanket_impls() {
    // Create independent build dir so all tests can run in parallel
    let build_dir = tempdir().unwrap();

    assert_public_api(
        builder_for_crate("../test-apis/example_api-v0.2.0", &build_dir).omit_blanket_impls(true),
        "example_api-v0.2.0-omit_blanket_impls",
    );
}

#[test]
fn omit_auto_trait_impls() {
    // Create independent build dir so all tests can run in parallel
    let build_dir = tempdir().unwrap();

    assert_public_api(
        builder_for_crate("../test-apis/example_api-v0.2.0", &build_dir)
            .omit_auto_trait_impls(true),
        "example_api-v0.2.0-omit_auto_trait_impls",
    );
}

#[test]
fn diff_with_added_items() {
    // Create independent build dirs so all tests can run in parallel
    let build_dir = tempdir().unwrap();
    let build_dir2 = tempdir().unwrap();

    assert_public_api_diff(
        rustdoc_json_path_for_crate("../test-apis/example_api-v0.1.0", &build_dir),
        rustdoc_json_path_for_crate("../test-apis/example_api-v0.2.0", &build_dir2),
        "diff_with_added_items",
    );
}

#[test]
fn empty_diff() {
    // Create independent build dirs so all tests can run in parallel
    let build_dir = tempdir().unwrap();
    let build_dir2 = tempdir().unwrap();

    // No change to the public API
    assert_public_api_diff(
        rustdoc_json_path_for_crate("../test-apis/comprehensive_api", &build_dir),
        rustdoc_json_path_for_crate("../test-apis/comprehensive_api", &build_dir2),
        "empty_diff",
    );
}

#[test]
fn diff_empty_when_item_moved_between_inherent_impls() {
    let v1 = rustdoc_json_for_lib(
        r#"
pub struct Foo;
impl Foo {
    pub fn f1() {}
    pub fn moved() {}
}
impl Foo {
    pub fn f2() {}
}
    "#,
    );

    let v2 = rustdoc_json_for_lib(
        r#"
pub struct Foo;
impl Foo {
    pub fn f1() {}
}
impl Foo {
    pub fn f2() {}
    pub fn moved() {}
}
    "#,
    );

    assert_public_api_diff(
        v1.json_path,
        v2.json_path,
        "diff_move_item_between_inherent_impls",
    );
}

#[test]
fn diff_empty_when_changing_inherent_impl_to_auto_derived_impl() {
    let v1 = rustdoc_json_for_lib(
        r#"
#[derive(Clone)]
pub enum Foo {
    A,
    B,
}

impl Default for Foo {
    fn default() -> Foo {
        Foo::A
    }
}
    "#,
    );

    let v2 = rustdoc_json_for_lib(
        r#"
#[derive(Clone, Default)]
pub enum Foo {
    #[default]
    A,
    B,
}
    "#,
    );

    // Note: The test will pass with assert_public_api_diff() because it uses
    // [`public_api::diff::PublicApiDiff::between`] which is smart enough to
    // realize there is no diff. But we also want to make sure the above case
    // does not result in any textual diff either.
    assert_no_textual_public_api_diff(v1.json_path, v2.json_path);
}

#[test]
fn diff_with_removed_items() {
    // Create independent build dirs so all tests can run in parallel
    let build_dir = tempdir().unwrap();
    let build_dir2 = tempdir().unwrap();

    assert_public_api_diff(
        rustdoc_json_path_for_crate("../test-apis/example_api-v0.2.0", &build_dir2),
        rustdoc_json_path_for_crate("../test-apis/example_api-v0.1.0", &build_dir),
        "diff_with_removed_items",
    );
}

#[test]
fn comprehensive_api() {
    // Create independent build dir so all tests can run in parallel
    let build_dir = tempdir().unwrap();

    assert_public_api(
        simplified_builder_for_crate("../test-apis/comprehensive_api", &build_dir),
        "comprehensive_api",
    );
}

#[test]
fn comprehensive_api_proc_macro() {
    // Create independent build dir so all tests can run in parallel
    let build_dir = tempdir().unwrap();

    assert_public_api(
        simplified_builder_for_crate("../test-apis/comprehensive_api_proc_macro", &build_dir),
        "comprehensive_api_proc_macro",
    );
}

#[test]
fn auto_traits() {
    // Create independent build dir so all tests can run in parallel
    let build_dir = tempdir().unwrap();

    assert_public_api(
        builder_for_crate("../test-apis/auto_traits", &build_dir).omit_blanket_impls(true),
        "auto_traits",
    );
}

/// Test that `debug_sorting` does not result in stack overflow because of
/// recursion. This can quite easily happen unless we test for it continuously.
/// We don't care what the exact output is, just that we don't crash.
#[test]
fn comprehensive_api_debug_sorting_no_stack_overflow() {
    // Create independent build dir so all tests can run in parallel
    let build_dir = tempdir().unwrap();

    let rustdoc_json = rustdoc_json_path_for_crate("../test-apis/comprehensive_api", &build_dir);
    let _api = public_api::Builder::from_rustdoc_json(rustdoc_json)
        .debug_sorting(true)
        .build()
        .unwrap()
        .to_string();
}

#[test]
fn invalid_json() {
    let invalid_json = NamedTempFile::new().unwrap();
    write!(invalid_json.as_file(), "}}}}}}}}}}").unwrap();
    let result = public_api::Builder::from_rustdoc_json(invalid_json.path()).build();
    assert!(matches!(result, Err(Error::SerdeJsonError(_))));
}

#[test]
fn find_function() {
    use rustdoc_types::{Crate, Id, ItemEnum};

    fn is_function(krate: &Crate, id: Id) -> bool {
        matches!(krate.index.get(&id).unwrap().inner, ItemEnum::Function(_))
    }

    let json = rustdoc_json_for_lib(
        r#"
pub mod a_mod {
    pub fn a_function() {
    }
}
    "#,
    );

    let public_api = public_api::Builder::from_rustdoc_json(&json.json_path)
        .build()
        .unwrap();

    let file = fs::File::open(json.json_path).unwrap();
    let krate = serde_json::from_reader::<_, Crate>(file).unwrap();

    let public_item = public_api
        .items()
        .find(|public_item| is_function(&krate, public_item.id()))
        .unwrap();

    let id = public_item.id();
    let item = krate.index.get(&id).unwrap();
    assert_eq!(Some("a_function"), item.name.as_deref());

    let parent_id = public_item.parent_id().unwrap();
    let parent_item = krate.index.get(&parent_id).unwrap();
    assert!(matches!(parent_item.inner, ItemEnum::Module(_)));
    assert_eq!(Some("a_mod"), parent_item.name.as_deref());
}

struct LibWithJson {
    json_path: PathBuf,

    /// Keep the tempdir alive so the json file doesn't get deleted
    _root: TempDir,
}

fn rustdoc_json_for_lib(lib: &str) -> LibWithJson {
    let root = tempdir().unwrap();

    let write = |file: &str, content: &str| {
        let file_path = root.path().join(file);
        fs::write(file_path, content).unwrap();
    };

    write(
        "Cargo.toml",
        "\
        [package]\n\
        name = \"lib\"\n\
        version = \"0.1.0\"\n\
        edition = \"2021\"\n\
        [lib]\n\
        path = \"lib.rs\"\n\
        ",
    );

    write("lib.rs", lib);

    LibWithJson {
        json_path: rustdoc_json_path_for_temp_crate(&root),
        _root: root,
    }
}

fn assert_public_api_diff(
    old_json: impl Into<PathBuf>,
    new_json: impl Into<PathBuf>,
    test_name: &str,
) {
    let old = public_api::Builder::from_rustdoc_json(old_json)
        .build()
        .unwrap();
    let new = public_api::Builder::from_rustdoc_json(new_json)
        .build()
        .unwrap();

    let diff = public_api::diff::PublicApiDiff::between(old, new);
    insta::assert_debug_snapshot!(test_name, diff);
}

// PublicApiDiff::between() is smarter than a textual diff, but in some cases we
// still care about a textual diff. When we care about the textual diff, we use
// this function.
fn assert_no_textual_public_api_diff(old_json: impl Into<PathBuf>, new_json: impl Into<PathBuf>) {
    let old = public_api::Builder::from_rustdoc_json(old_json)
        .build()
        .unwrap()
        .to_string();
    let new = public_api::Builder::from_rustdoc_json(new_json)
        .build()
        .unwrap()
        .to_string();

    assert_eq!(new, old);
}

/// Asserts that the public API of the crate in the given rustdoc JSON file
/// matches the expected output.
fn assert_public_api(builder: public_api::Builder, test_name: &str) {
    let api = builder.build().unwrap().to_string();

    insta::assert_snapshot!(test_name, api);
}
