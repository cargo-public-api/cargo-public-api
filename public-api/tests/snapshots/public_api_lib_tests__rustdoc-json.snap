---
source: public-api/tests/public-api-lib-tests.rs
expression: public_api
---
pub mod rustdoc_json
#[non_exhaustive] pub enum rustdoc_json::BuildError
pub rustdoc_json::BuildError::BuildRustdocJsonError
pub rustdoc_json::BuildError::CapturedOutputError(alloc::string::String)
pub rustdoc_json::BuildError::CargoManifestError(cargo_manifest::error::Error)
pub rustdoc_json::BuildError::CargoMetadataError(cargo_metadata::errors::Error)
pub rustdoc_json::BuildError::CommandExecutionError(alloc::string::String)
pub rustdoc_json::BuildError::General(alloc::string::String)
pub rustdoc_json::BuildError::IoError(std::io::error::Error)
pub rustdoc_json::BuildError::VirtualManifest(std::path::PathBuf)
impl core::convert::From<cargo_manifest::error::Error> for rustdoc_json::BuildError
pub fn rustdoc_json::BuildError::from(source: cargo_manifest::error::Error) -> Self
impl core::convert::From<cargo_metadata::errors::Error> for rustdoc_json::BuildError
pub fn rustdoc_json::BuildError::from(source: cargo_metadata::errors::Error) -> Self
impl core::convert::From<std::io::error::Error> for rustdoc_json::BuildError
pub fn rustdoc_json::BuildError::from(source: std::io::error::Error) -> Self
impl core::error::Error for rustdoc_json::BuildError
pub fn rustdoc_json::BuildError::source(&self) -> core::option::Option<&(dyn core::error::Error + 'static)>
impl core::fmt::Debug for rustdoc_json::BuildError
pub fn rustdoc_json::BuildError::fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result
impl core::fmt::Display for rustdoc_json::BuildError
pub fn rustdoc_json::BuildError::fmt(&self, __formatter: &mut core::fmt::Formatter<'_>) -> core::fmt::Result
impl core::marker::Freeze for rustdoc_json::BuildError
impl core::marker::Send for rustdoc_json::BuildError
impl core::marker::Sync for rustdoc_json::BuildError
impl core::marker::Unpin for rustdoc_json::BuildError
impl !core::panic::unwind_safe::RefUnwindSafe for rustdoc_json::BuildError
impl !core::panic::unwind_safe::UnwindSafe for rustdoc_json::BuildError
impl<T, U> core::convert::Into<U> for rustdoc_json::BuildError where U: core::convert::From<T>
pub fn rustdoc_json::BuildError::into(self) -> U
impl<T, U> core::convert::TryFrom<U> for rustdoc_json::BuildError where U: core::convert::Into<T>
pub type rustdoc_json::BuildError::Error = core::convert::Infallible
pub fn rustdoc_json::BuildError::try_from(value: U) -> core::result::Result<T, <T as core::convert::TryFrom<U>>::Error>
impl<T, U> core::convert::TryInto<U> for rustdoc_json::BuildError where U: core::convert::TryFrom<T>
pub type rustdoc_json::BuildError::Error = <U as core::convert::TryFrom<T>>::Error
pub fn rustdoc_json::BuildError::try_into(self) -> core::result::Result<U, <U as core::convert::TryFrom<T>>::Error>
impl<T> alloc::string::ToString for rustdoc_json::BuildError where T: core::fmt::Display + ?core::marker::Sized
pub fn rustdoc_json::BuildError::to_string(&self) -> alloc::string::String
impl<T> core::any::Any for rustdoc_json::BuildError where T: 'static + ?core::marker::Sized
pub fn rustdoc_json::BuildError::type_id(&self) -> core::any::TypeId
impl<T> core::borrow::Borrow<T> for rustdoc_json::BuildError where T: ?core::marker::Sized
pub fn rustdoc_json::BuildError::borrow(&self) -> &T
impl<T> core::borrow::BorrowMut<T> for rustdoc_json::BuildError where T: ?core::marker::Sized
pub fn rustdoc_json::BuildError::borrow_mut(&mut self) -> &mut T
impl<T> core::convert::From<T> for rustdoc_json::BuildError
pub fn rustdoc_json::BuildError::from(t: T) -> T
impl<T> tracing::instrument::Instrument for rustdoc_json::BuildError
impl<T> tracing::instrument::WithSubscriber for rustdoc_json::BuildError
pub enum rustdoc_json::Color
pub rustdoc_json::Color::Always
pub rustdoc_json::Color::Auto
pub rustdoc_json::Color::Never
impl core::clone::Clone for rustdoc_json::Color
pub fn rustdoc_json::Color::clone(&self) -> rustdoc_json::Color
impl core::fmt::Debug for rustdoc_json::Color
pub fn rustdoc_json::Color::fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result
impl core::marker::Copy for rustdoc_json::Color
impl core::marker::Freeze for rustdoc_json::Color
impl core::marker::Send for rustdoc_json::Color
impl core::marker::Sync for rustdoc_json::Color
impl core::marker::Unpin for rustdoc_json::Color
impl core::panic::unwind_safe::RefUnwindSafe for rustdoc_json::Color
impl core::panic::unwind_safe::UnwindSafe for rustdoc_json::Color
impl<T, U> core::convert::Into<U> for rustdoc_json::Color where U: core::convert::From<T>
pub fn rustdoc_json::Color::into(self) -> U
impl<T, U> core::convert::TryFrom<U> for rustdoc_json::Color where U: core::convert::Into<T>
pub type rustdoc_json::Color::Error = core::convert::Infallible
pub fn rustdoc_json::Color::try_from(value: U) -> core::result::Result<T, <T as core::convert::TryFrom<U>>::Error>
impl<T, U> core::convert::TryInto<U> for rustdoc_json::Color where U: core::convert::TryFrom<T>
pub type rustdoc_json::Color::Error = <U as core::convert::TryFrom<T>>::Error
pub fn rustdoc_json::Color::try_into(self) -> core::result::Result<U, <U as core::convert::TryFrom<T>>::Error>
impl<T> alloc::borrow::ToOwned for rustdoc_json::Color where T: core::clone::Clone
pub type rustdoc_json::Color::Owned = T
pub fn rustdoc_json::Color::clone_into(&self, target: &mut T)
pub fn rustdoc_json::Color::to_owned(&self) -> T
impl<T> core::any::Any for rustdoc_json::Color where T: 'static + ?core::marker::Sized
pub fn rustdoc_json::Color::type_id(&self) -> core::any::TypeId
impl<T> core::borrow::Borrow<T> for rustdoc_json::Color where T: ?core::marker::Sized
pub fn rustdoc_json::Color::borrow(&self) -> &T
impl<T> core::borrow::BorrowMut<T> for rustdoc_json::Color where T: ?core::marker::Sized
pub fn rustdoc_json::Color::borrow_mut(&mut self) -> &mut T
impl<T> core::clone::CloneToUninit for rustdoc_json::Color where T: core::clone::Clone
pub unsafe fn rustdoc_json::Color::clone_to_uninit(&self, dest: *mut u8)
impl<T> core::convert::From<T> for rustdoc_json::Color
pub fn rustdoc_json::Color::from(t: T) -> T
impl<T> tracing::instrument::Instrument for rustdoc_json::Color
impl<T> tracing::instrument::WithSubscriber for rustdoc_json::Color
#[non_exhaustive] pub enum rustdoc_json::PackageTarget
pub rustdoc_json::PackageTarget::Bench(alloc::string::String)
pub rustdoc_json::PackageTarget::Bin(alloc::string::String)
pub rustdoc_json::PackageTarget::Example(alloc::string::String)
pub rustdoc_json::PackageTarget::Lib
pub rustdoc_json::PackageTarget::Test(alloc::string::String)
impl core::clone::Clone for rustdoc_json::PackageTarget
pub fn rustdoc_json::PackageTarget::clone(&self) -> rustdoc_json::PackageTarget
impl core::default::Default for rustdoc_json::PackageTarget
pub fn rustdoc_json::PackageTarget::default() -> rustdoc_json::PackageTarget
impl core::fmt::Debug for rustdoc_json::PackageTarget
pub fn rustdoc_json::PackageTarget::fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result
impl core::marker::Freeze for rustdoc_json::PackageTarget
impl core::marker::Send for rustdoc_json::PackageTarget
impl core::marker::Sync for rustdoc_json::PackageTarget
impl core::marker::Unpin for rustdoc_json::PackageTarget
impl core::panic::unwind_safe::RefUnwindSafe for rustdoc_json::PackageTarget
impl core::panic::unwind_safe::UnwindSafe for rustdoc_json::PackageTarget
impl<T, U> core::convert::Into<U> for rustdoc_json::PackageTarget where U: core::convert::From<T>
pub fn rustdoc_json::PackageTarget::into(self) -> U
impl<T, U> core::convert::TryFrom<U> for rustdoc_json::PackageTarget where U: core::convert::Into<T>
pub type rustdoc_json::PackageTarget::Error = core::convert::Infallible
pub fn rustdoc_json::PackageTarget::try_from(value: U) -> core::result::Result<T, <T as core::convert::TryFrom<U>>::Error>
impl<T, U> core::convert::TryInto<U> for rustdoc_json::PackageTarget where U: core::convert::TryFrom<T>
pub type rustdoc_json::PackageTarget::Error = <U as core::convert::TryFrom<T>>::Error
pub fn rustdoc_json::PackageTarget::try_into(self) -> core::result::Result<U, <U as core::convert::TryFrom<T>>::Error>
impl<T> alloc::borrow::ToOwned for rustdoc_json::PackageTarget where T: core::clone::Clone
pub type rustdoc_json::PackageTarget::Owned = T
pub fn rustdoc_json::PackageTarget::clone_into(&self, target: &mut T)
pub fn rustdoc_json::PackageTarget::to_owned(&self) -> T
impl<T> core::any::Any for rustdoc_json::PackageTarget where T: 'static + ?core::marker::Sized
pub fn rustdoc_json::PackageTarget::type_id(&self) -> core::any::TypeId
impl<T> core::borrow::Borrow<T> for rustdoc_json::PackageTarget where T: ?core::marker::Sized
pub fn rustdoc_json::PackageTarget::borrow(&self) -> &T
impl<T> core::borrow::BorrowMut<T> for rustdoc_json::PackageTarget where T: ?core::marker::Sized
pub fn rustdoc_json::PackageTarget::borrow_mut(&mut self) -> &mut T
impl<T> core::clone::CloneToUninit for rustdoc_json::PackageTarget where T: core::clone::Clone
pub unsafe fn rustdoc_json::PackageTarget::clone_to_uninit(&self, dest: *mut u8)
impl<T> core::convert::From<T> for rustdoc_json::PackageTarget
pub fn rustdoc_json::PackageTarget::from(t: T) -> T
impl<T> tracing::instrument::Instrument for rustdoc_json::PackageTarget
impl<T> tracing::instrument::WithSubscriber for rustdoc_json::PackageTarget
pub struct rustdoc_json::Builder
impl rustdoc_json::Builder
pub const fn rustdoc_json::Builder::all_features(self, all_features: bool) -> Self
pub fn rustdoc_json::Builder::build(self) -> core::result::Result<std::path::PathBuf, rustdoc_json::BuildError>
pub fn rustdoc_json::Builder::build_with_captured_output(self, stdout: impl std::io::Write, stderr: impl std::io::Write) -> core::result::Result<std::path::PathBuf, rustdoc_json::BuildError>
pub fn rustdoc_json::Builder::cap_lints(self, cap_lints: core::option::Option<impl core::convert::AsRef<str>>) -> Self
pub fn rustdoc_json::Builder::clear_target_dir(self) -> Self
pub fn rustdoc_json::Builder::clear_toolchain(self) -> Self
pub const fn rustdoc_json::Builder::color(self, color: rustdoc_json::Color) -> Self
pub fn rustdoc_json::Builder::document_private_items(self, document_private_items: bool) -> Self
pub fn rustdoc_json::Builder::features<I: core::iter::traits::collect::IntoIterator<Item = S>, S: core::convert::AsRef<str>>(self, features: I) -> Self
pub fn rustdoc_json::Builder::manifest_path(self, manifest_path: impl core::convert::AsRef<std::path::Path>) -> Self
pub const fn rustdoc_json::Builder::no_default_features(self, no_default_features: bool) -> Self
pub fn rustdoc_json::Builder::package(self, package: impl core::convert::AsRef<str>) -> Self
pub fn rustdoc_json::Builder::package_target(self, package_target: rustdoc_json::PackageTarget) -> Self
pub const fn rustdoc_json::Builder::quiet(self, quiet: bool) -> Self
pub const fn rustdoc_json::Builder::silent(self, silent: bool) -> Self
pub fn rustdoc_json::Builder::target(self, target: alloc::string::String) -> Self
pub fn rustdoc_json::Builder::target_dir(self, target_dir: impl core::convert::AsRef<std::path::Path>) -> Self
pub fn rustdoc_json::Builder::toolchain(self, toolchain: impl core::convert::Into<alloc::string::String>) -> Self
impl core::clone::Clone for rustdoc_json::Builder
pub fn rustdoc_json::Builder::clone(&self) -> rustdoc_json::Builder
impl core::default::Default for rustdoc_json::Builder
pub fn rustdoc_json::Builder::default() -> Self
impl core::fmt::Debug for rustdoc_json::Builder
pub fn rustdoc_json::Builder::fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result
impl core::marker::Freeze for rustdoc_json::Builder
impl core::marker::Send for rustdoc_json::Builder
impl core::marker::Sync for rustdoc_json::Builder
impl core::marker::Unpin for rustdoc_json::Builder
impl core::panic::unwind_safe::RefUnwindSafe for rustdoc_json::Builder
impl core::panic::unwind_safe::UnwindSafe for rustdoc_json::Builder
impl<T, U> core::convert::Into<U> for rustdoc_json::Builder where U: core::convert::From<T>
pub fn rustdoc_json::Builder::into(self) -> U
impl<T, U> core::convert::TryFrom<U> for rustdoc_json::Builder where U: core::convert::Into<T>
pub type rustdoc_json::Builder::Error = core::convert::Infallible
pub fn rustdoc_json::Builder::try_from(value: U) -> core::result::Result<T, <T as core::convert::TryFrom<U>>::Error>
impl<T, U> core::convert::TryInto<U> for rustdoc_json::Builder where U: core::convert::TryFrom<T>
pub type rustdoc_json::Builder::Error = <U as core::convert::TryFrom<T>>::Error
pub fn rustdoc_json::Builder::try_into(self) -> core::result::Result<U, <U as core::convert::TryFrom<T>>::Error>
impl<T> alloc::borrow::ToOwned for rustdoc_json::Builder where T: core::clone::Clone
pub type rustdoc_json::Builder::Owned = T
pub fn rustdoc_json::Builder::clone_into(&self, target: &mut T)
pub fn rustdoc_json::Builder::to_owned(&self) -> T
impl<T> core::any::Any for rustdoc_json::Builder where T: 'static + ?core::marker::Sized
pub fn rustdoc_json::Builder::type_id(&self) -> core::any::TypeId
impl<T> core::borrow::Borrow<T> for rustdoc_json::Builder where T: ?core::marker::Sized
pub fn rustdoc_json::Builder::borrow(&self) -> &T
impl<T> core::borrow::BorrowMut<T> for rustdoc_json::Builder where T: ?core::marker::Sized
pub fn rustdoc_json::Builder::borrow_mut(&mut self) -> &mut T
impl<T> core::clone::CloneToUninit for rustdoc_json::Builder where T: core::clone::Clone
pub unsafe fn rustdoc_json::Builder::clone_to_uninit(&self, dest: *mut u8)
impl<T> core::convert::From<T> for rustdoc_json::Builder
pub fn rustdoc_json::Builder::from(t: T) -> T
impl<T> tracing::instrument::Instrument for rustdoc_json::Builder
impl<T> tracing::instrument::WithSubscriber for rustdoc_json::Builder
